File: web/src/commands/links.ts
import type { Provider } from '../api/vocabulary'
import type { CrosstuneDb } from '../db/schema'
import { newId, nextPosition, now, putRow, tombstone, writeTx } from './write'

export interface LinkInput {
  url: string
  provider: Provider
  provider_ref?: string | null
  title?: string | null
  artwork_url?: string | null
  label?: string | null
}

export async function addLink(db: CrosstuneDb, songId: string, link: LinkInput): Promise<string> {
  const at = now()
  const id = newId()
  await writeTx(db, async () => {
    const song = await db.songs.get(songId)
    if (!song || song.deleted_at) throw new Error('Song not found')
    const existing = await db.recording_links.where('song_id').equals(songId).toArray()
    const active = existing.filter((l) => !l.deleted_at)
    await putRow(db, 'recording_links', {
      id,
      created_at: at,
      updated_at: at,
      deleted_at: null,
      server_seq: 0,
      song_id: songId,
      url: link.url.trim(),
      provider: link.provider,
      provider_ref: link.provider_ref ?? null,
      title: link.title ?? null,
      artwork_url: link.artwork_url ?? null,
      label: link.label ?? null,
      position: nextPosition(active),
    })
  })
  return id
}

export async function removeLink(db: CrosstuneDb, linkId: string): Promise<void> {
  await writeTx(db, () => tombstone(db, 'recording_links', linkId, now()))
}

File: apple/CrosstuneKit/Sources/CrosstuneCommands/Links.swift
import CrosstuneStore
import Foundation
import GRDB

/// What a screen gathers to attach a recording link to a tune.
public struct LinkInput: Sendable {
    public var url: String
    public var provider: String
    public var providerRef: String?
    public var title: String?
    public var artworkURL: String?
    public var label: String?

    public init(
        url: String, provider: String, providerRef: String? = nil, title: String? = nil,
        artworkURL: String? = nil, label: String? = nil
    ) {
        self.url = url
        self.provider = provider
        self.providerRef = providerRef
        self.title = title
        self.artworkURL = artworkURL
        self.label = label
    }
}

extension StoreWriter {
    /// Appends a link to a live tune's list, after its currently active links.
    @discardableResult
    public func addLink(tuneID: String, link: LinkInput, at time: Timestamp = .now) throws -> String {
        guard let tune = try Tune.fetchOne(db, key: tuneID), tune.deletedAt == nil else {
            throw CommandError.tuneNotFound
        }
        let active = try RecordingLink.filter(Column("tune_id") == tuneID).fetchAll(db).filter { $0.deletedAt == nil }
        let id = newID(at: time)
        try put(
            RecordingLink(
                id: id, createdAt: time, updatedAt: time, tuneID: tuneID,
                url: link.url.trimmingCharacters(in: .whitespacesAndNewlines), provider: link.provider,
                providerRef: link.providerRef, title: link.title, label: link.label,
                artworkURL: link.artworkURL, position: nextPosition(active)), at: time)
        return id
    }

    public func removeLink(_ linkID: String, at time: Timestamp = .now) throws {
        try tombstone(RecordingLink.self, id: linkID, at: time)
    }
}

extension Commands {
    /// Appends a link to a tune's list.
    @discardableResult
    public func addLink(tuneID: String, link: LinkInput, at time: Timestamp = .now) async throws -> String {
        try await store.write { writer in try writer.addLink(tuneID: tuneID, link: link, at: time) }
    }

    public func removeLink(_ linkID: String, at time: Timestamp = .now) async throws {
        try await store.write { writer in try writer.removeLink(linkID, at: time) }
    }
}
